### PR TITLE
Allow to declare (instrumentation) dependencies

### LINKED ISSUE #3983: Is it possible to specify dependencies for an instrumenter?
Is it possible to specify dependencies for an instrumenter? For example, in Bisect, we previously recommended using `(preprocessor_deps ...)` to give the filename of the exclusions file, i.e.

```
(preprocess (pps bisect_ppx --exclusions bisect.exclude))
(preprocessor_deps (file bisect.exclude))
```

Doing this now with

```
(instrumentation (backend bisect_ppx --exclusions bisect.exclude))
(preprocessor_deps (file bisect.exclude))
```

results in

```
5 |  (preprocessor_deps (file bisect.exclude))
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This preprocessor_deps field will be ignored because no preprocessor
that might use them is configured.
```

_Originally posted by @aantron in https://github.com/ocaml/dune/issues/3932#issuecomment-734738991_

### PR DESCRIPTION
**Pull request by nojb**
Fixes #3983 

@aantron if you could give this one a spin that would be great. The syntax is as discussed in #3983.

### CODE REVIEW

**Comment by aantron** [comment]
Looks good and seems to work.

File: src/dune_rules/lib_rules.ml [hunk 1]
```diff
@@ -344,15 +344,21 @@ let cctx (lib : Library.t) ~sctx ~source_modules ~dir ~expander ~scope
...
+  in
+  let instrumentation_deps =
+    Preprocess.Per_module.instrumentation_deps lib.buildable.preprocess
+      ~instrumentation_backend
   in
   (* Preprocess before adding the alias module as it doesn't need preprocessing *)
   let pp =
     Preprocessing.make sctx ~dir ~dep_kind ~scope ~preprocess ~expander
-      ~preprocessor_deps:lib.buildable.preprocessor_deps
+      ~preprocessor_deps:(lib.buildable.preprocessor_deps @ instrumentation_deps)
```
Comment (jeremiedimino) [inline src/dune_rules/lib_rules.ml:361]: It might be worth sharing this code betweem `exe_rules.ml` and `lib_rules.ml`.

**Review by jeremiedimino — APPROVED** [review]
LGTM.

File: src/dune_rules/lib_rules.ml (hunk 1 above)
Comment (nojb) [reply src/dune_rules/lib_rules.ml]: Good idea. I added an extra argument to `Preprocessing.make ~instrumentation_deps` and do the concatentation inside that function.

**Comment by nojb** [comment]
Will merge this once CI is green.

@rgrinberg this should go into 2.8.3... to this end, should I change the version guard in the code from `(3, 0)` to `(2, 8)` ?

**Comment by rgrinberg** [comment]
This is a new feature, so it cannot go into 2.8.3

**Comment by nojb** [comment]
> This is a new feature, so it cannot go into 2.8.3

Right, sorry, I got confused for a moment :)

**Comment by rgrinberg** [comment]
If you'd like we could include it in 2.9 however. The PR will need to be modified slightly though.

**Comment by ejgallego** [comment]
Hi folks, deadline for submitting the backport of this PR is June 11th, please do so or remove the 2.9 milestone from this PR if you don't want this in 2.9 anymore.

**Comment by nojb** [comment]
PR backport at #4686

**Comment by ejgallego** [comment]
> PR backport at #4686

Thanks a lot @nojb !

**Pull request merged**